tests/storages_tests/test_storages.py

```diff
@@ -352,336 +452,406 @@ def test_get_trial_number_from_id(storage_mode):
+@pytest.mark.parametrize("storage_mode", STORAGE_MODES)
+def test_set_trial_param(storage_mode: str) -> None:
+
+    with StorageSupplier(storage_mode) as storage:
+
+        # Setup test across multiple studies and trials.
+        study_id = storage.create_new_study()
+        trial_id_1 = storage.create_new_trial(study_id)
+        trial_id_2 = storage.create_new_trial(study_id)
+        trial_id_3 = storage.create_new_trial(storage.create_new_study())
+
+        # Setup Distributions.
+        distribution_x = UniformDistribution(low=1.0, high=2.0)
+        distribution_y_1 = CategoricalDistribution(choices=("Shibuya", "Ebisu", "Meguro"))
+        distribution_y_2 = CategoricalDistribution(choices=("Shibuya", "Shinsen"))
+        distribution_z = LogUniformDistribution(low=1.0, high=100.0)
+
+        # Set new params.
+        assert storage.set_trial_param(trial_id_1, "x", 0.5, distribution_x)
+        assert storage.set_trial_param(trial_id_1, "y", 2, distribution_y_1)
+        assert storage.get_trial_param(trial_id_1, "x") == 0.5
+        assert storage.get_trial_param(trial_id_1, "y") == 2
+        # Check set_param breaks neither get_trial nor get_trial_params
+        assert storage.get_trial(trial_id_1).params == {"x": 0.5, "y": "Meguro"}
+        assert storage.get_trial_params(trial_id_1) == {"x": 0.5, "y": "Meguro"}
+        # Duplicated registration should return False.
+        assert not storage.set_trial_param(trial_id_1, "x", 0.6, distribution_x)
+        # Duplicated registration should not change the existing value.
+        assert storage.get_trial_param(trial_id_1, "x") == 0.5
+
+        # Set params to another trial.
+        assert storage.set_trial_param(trial_id_2, "x", 0.3, distribution_x)
+        assert storage.set_trial_param(trial_id_2, "z", 0.1, distribution_z)
+        assert storage.get_trial_param(trial_id_2, "x") == 0.3
+        assert storage.get_trial_param(trial_id_2, "z") == 0.1
+        assert storage.get_trial(trial_id_2).params == {"x": 0.3, "z": 0.1}
+        assert storage.get_trial_params(trial_id_2) == {"x": 0.3, "z": 0.1}
+
+        # Set params with distributions that does not match previous ones.
+        with pytest.raises(ValueError):
+            storage.set_trial_param(trial_id_2, "x", 0.5, distribution_z)
+        with pytest.raises(ValueError):
+            storage.set_trial_param(trial_id_2, "y", 0.5, distribution_z)
+        # Choices in CategoricalDistribution should match including its order.
+        with pytest.raises(ValueError):
+            storage.set_trial_param(
+                trial_id_2, "y", 2, CategoricalDistribution(choices=("Meguro", "Shibuya", "Ebisu"))
+            )
+
+        storage.set_trial_state(trial_id_2, TrialState.COMPLETE)
+        # Cannot assign params to finished trial.
+        with pytest.raises(RuntimeError):
+            storage.set_trial_param(trial_id_2, "y", 2, distribution_y_1)
+        # Check the previous call does not change the trial's state
```

Comment:
IMO, `trial's state` is confusing with the `TrialState`. How about replacing it with `params`?

```suggestion
        # Check the previous call does not change the params.
```

Reply:
I agree with the comment and commit the suggestion.